django/db/models/base.py

```diff
@@ -1287,10 +1287,30 @@ def check(cls, **kwargs):
                 *cls._check_indexes(databases),
                 *cls._check_ordering(),
                 *cls._check_constraints(databases),
+                *cls._check_default_pk(),
             ]
 
         return errors
 
+    @classmethod
+    def _check_default_pk(cls):
+        app_config = cls._meta.app_config
+        overridden_pk_appconfig = app_config._is_model_default_pk_overridden()
+        auto_created_pk = cls._meta.pk.auto_created
+        overridden_pk_setting = settings.is_overridden("DEFAULT_MODEL_PK")
+        if auto_created_pk and not overridden_pk_setting and not overridden_pk_appconfig:
+            return [
+                checks.Error(
+                    "Auto-created primary key used without defining a primary key type",
+                    hint=f"Set the DEFAULT_MODEL_PK setting or {app_config.__class__.__name__}.default_auto_field "
+                         f"attribute to either django.db.models.BigAutoField' or the legacy "
+                         f"'django.db.models.AutoField'.",
```

Comment:
I wouldn't call `AutoField` a "legacy", we should also include `SmallAutoField`.

Reply:
Is `SmallAutoField` ever the right choice for a primary key field? I really think not. The idea here is to guide people towards the right choice which is almost always `BigAutoField`. For completeness I've included it, but I think that it might be a mistake.

Reply:
> Is SmallAutoField ever the right choice for a primary key field?

Why not, it can the right choice for any small dictionary tables if you really want to save space. I would even say that it's justified more often then `BigAutoField` :wink:

Reply:
In one specific case it might be correct, and that would be: a small dictionary table, in an environment where saving between 2 and 4 bytes per row is important, and you are 100% sure you won't need to grow this table ever.

I think this is really specific enough to not warrant it appearing in a general error/warning message. Someone might look at this and say "I didn't know SmallAutoField exists, my data is small so it would be perfect!", which conflicts with the spirit of this work, which was to help people use a more future proof primary key field.